### PR TITLE
fix: scalar timestamp assignment (#19843)

### LINKED ISSUE #19843: BUG: scalar assignment of a tz-aware is object dtype
[3] should be a ``datetime64[ns, UTC]``

```
In [1]: df = pd.DataFrame({'A': [0, 1]})

In [3]: df['now'] = pd.Timestamp('20130101', tz='UTC')

In [4]: df
Out[4]: 
   A                        now
0  0  2013-01-01 00:00:00+00:00
1  1  2013-01-01 00:00:00+00:00

In [5]: df.dtypes
Out[5]: 
A       int64
now    object
dtype: object

In [6]: df['now2'] = pd.DatetimeIndex([pd.Timestamp('20130101', tz='UTC')]).repeat(len(df))

In [7]: df.dtypes
Out[7]: 
A                     int64
now                  object
now2    datetime64[ns, UTC]
dtype: object
```

### PR DESCRIPTION
**Pull request by DylanDmitri**
- [x] closes #19843
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/scalar/timestamp/test_assignment_dtype.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import pandas as pd
+from pandas.core.dtypes.dtypes import DatetimeTZDtype
+
+
+# referencing #19843: scalar assignment of a tz-aware is object dtype
```
Comment (gfyoung) [inline pandas/tests/scalar/timestamp/test_assignment_dtype.py]: I think we can find a home for this in another test file (and don't need a new one).  In addition, the comment should be under the class declaration.

@jreback : I think `tests/frame/test_indexing.py` makes sense here.  What do you think?

Comment (DylanDmitri) [reply pandas/tests/scalar/timestamp/test_assignment_dtype.py]: Sounds good, should they go under ```TestDataFrameIndexingDatetimeWithTZ```, or in their own class?

Comment (gfyoung) [reply pandas/tests/scalar/timestamp/test_assignment_dtype.py]: Same class sounds good for now.

File: pandas/tests/frame/test_indexing.py [hunk 2]
```diff
@@ -3042,6 +3044,22 @@ def test_transpose(self):
         expected.index = ['A', 'B']
         assert_frame_equal(result, expected)
 
+    def test_scalar_assignment(self):
```
Comment (gfyoung) [inline pandas/tests/frame/test_indexing.py]: Reference issue number for both tests.

**Comment by DylanDmitri** [comment]
what should I do for the `whatsnew`?

File: doc/source/whatsnew/v0.23.0.txt [hunk 3]
```diff
@@ -854,6 +854,7 @@ Datetimelike
 - Bug in :func:`to_datetime` where passing an out-of-bounds datetime with ``errors='coerce'`` and ``utc=True`` would raise ``OutOfBoundsDatetime`` instead of parsing to ``NaT`` (:issue:`19612`)
 - Bug in :class:`DatetimeIndex` and :class:`TimedeltaIndex` addition and subtraction where name of the returned object was not always set consistently. (:issue:`19744`)
 - Bug in :class:`DatetimeIndex` and :class:`TimedeltaIndex` addition and subtraction where operations with numpy arrays raised ``TypeError`` (:issue:`19847`)
+- Bug in `DataFrame` assignment with a timezone-aware object (:issue:`19843`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: use double-backticks are DataFrame

say a timezone-aware scalar

File: pandas/core/frame.py [hunk 4]
```diff
@@ -2868,9 +2868,15 @@ def reindexer(value):
                 value = maybe_infer_to_datetimelike(value)
 
         else:
-            # upcast the scalar
+            # cast ignores pandas dtypes. so save the dtype first
+            from pandas.core.dtypes.cast import infer_dtype_from_scalar
```
Comment (jreback) [inline pandas/core/frame.py]: import at the top of the file with other imports

File: pandas/core/frame.py [hunk 5]
```diff
@@ -2868,9 +2868,15 @@ def reindexer(value):
...
+            pd_dtype, _ = infer_dtype_from_scalar(value, pandas_dtype=True)
```
Comment (jreback) [inline pandas/core/frame.py]: call the return dtype

File: pandas/core/frame.py [hunk 6]
```diff
@@ -2868,9 +2868,15 @@ def reindexer(value):
...
+
+            # upcast
             value = cast_scalar_to_array(len(self.index), value)
-            value = maybe_cast_to_datetime(value, value.dtype)
+
```
Comment (jreback) [inline pandas/core/frame.py]: you don't need the blank line and the comment here

File: pandas/tests/frame/test_indexing.py [hunk 7]
```diff
@@ -3042,6 +3044,20 @@ def test_transpose(self):
...
+        # issue #19843
+        df = pd.DataFrame(index=(0, 1, 2))
+        df['now'] = pd.Timestamp('20130101', tz='UTC')
+        assert isinstance(df.dtypes[0], DatetimeTZDtype)
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: construct the expected frame and use ``assert_frame_equal``

File: pandas/tests/frame/test_indexing.py [hunk 8]
```diff
@@ -3042,6 +3044,20 @@ def test_transpose(self):
...
+
+    def test_datetime_index_assignment(self):
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: this is a duplicate test.

File: pandas/tests/frame/test_indexing.py (hunk 7 above)
Comment (jreback) [reply pandas/tests/frame/test_indexing.py]: put this in ``class TestDataFrameIndexingDatetimeWithTZ(TestData)``

name it ``test_setitem_scalar``

parameterize it with both a tz and a tz-aware value.

File: pandas/tests/frame/test_indexing.py (hunk 8 above)
Comment (jreback) [reply pandas/tests/frame/test_indexing.py]: you could put it as an additional test in the class as above if you'd like (parametrize as well)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you rebase and update

**Comment by jreback** [comment]
can you rebase & move the note to 0.24.0

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
fixed up, but going to do in 0.23.4.

**Pull request merged**